### PR TITLE
[FEATURE] Compression and decompression for structure files

### LINKED ISSUE #645: (De)Compression support for structure_file
Basically just do #626 but for structure files.

### PR DESCRIPTION
**Pull request by joergi-w**
resolves #645

### CODE REVIEW

**Comment by joergi-w** [comment]
Yes, it is on current master and the concepts have not been not renamed yet.

File: include/seqan3/io/structure_file/format_vienna.hpp [hunk 1]
```diff
@@ -129,9 +129,10 @@ class structure_file_format_vienna
               comment_type & SEQAN3_DOXYGEN_ONLY(comment),
               offset_type & SEQAN3_DOXYGEN_ONLY(offset))
     {
-        auto stream_view = view::subrange<decltype(std::istreambuf_iterator<char>{stream}),
-                                          decltype(std::istreambuf_iterator<char>{})>
-                           { std::istreambuf_iterator<char>{stream}, std::istreambuf_iterator<char>{} };
+        using stream_char_t = typename stream_type::char_type;
```
Comment (smehringer) [inline include/seqan3/io/structure_file/format_vienna.hpp]:
```suggestion
        using stream_t = typename std::istreambuf_iterator<typename stream_type::char_type>;
```
make the lines below shorter and also line 259

File: include/seqan3/io/structure_file/input.hpp [hunk 2]
```diff
@@ -856,19 +845,48 @@ class structure_file_in
...
+        read_next_record();
+    }
+
+    //!\overload
+    template<istream_concept2 stream_t, structure_file_input_format_concept file_format>
+    structure_file_in(stream_t && stream,
+                      file_format const & SEQAN3_DOXYGEN_ONLY(format_tag),
+                      selected_field_ids const & SEQAN3_DOXYGEN_ONLY(fields_tag) = selected_field_ids{}) :
+        primary_stream{new stream_t{std::move(stream)}, stream_deleter_default},
+        format{file_format{}}
```
Comment (smehringer) [inline include/seqan3/io/structure_file/input.hpp]: can go into line above

File: include/seqan3/io/structure_file/output.hpp [hunk 3]
```diff
@@ -332,38 +332,70 @@ class structure_file_out
...
         // initialise format handler or throw if format is not found
-        detail::set_format(format, _file_name);
+        detail::set_format(format, filename);
     }
 
     /*!\brief Construct from an existing stream and with specified format.
      * \tparam file_format The format of the file in the stream, must satisfy
      * seqan3::structure_file_output_format_concept.
-     * \param[in] _stream  The stream to operate on (this must be std::move'd in!).
+     * \param[in,out] stream  The stream to write to, must be derived of std::basic_ostream<stream_char_t>.
```
Comment (smehringer) [inline include/seqan3/io/structure_file/output.hpp]:
```suggestion
     * \param[in,out] stream  The stream to write to, must be derived of std::basic_ostream.
```
`stream_char_t` is not defined in this function

File: test/unit/io/structure_file/structure_file_output_test.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
 // shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
 // -----------------------------------------------------------------------------------------------------
 
+#include <cstdio>
```
Comment (smehringer) [inline test/unit/io/structure_file/structure_file_output_test.cpp]: for what do you need this include?

File: include/seqan3/io/structure_file/format_vienna.hpp (hunk 1 above)
Comment (joergi-w) [reply include/seqan3/io/structure_file/format_vienna.hpp]: True! That's a really good idea.

File: include/seqan3/io/structure_file/output.hpp (hunk 3 above)
Comment (joergi-w) [reply include/seqan3/io/structure_file/output.hpp]: ok!

File: test/unit/io/structure_file/structure_file_output_test.cpp (hunk 4 above)
Comment (joergi-w) [reply test/unit/io/structure_file/structure_file_output_test.cpp]: I had used std::printf before, but deleted it, so the header is not needed anymore. Well spotted!

File: include/seqan3/io/structure_file/format_vienna.hpp [hunk 5]
```diff
@@ -255,7 +255,8 @@ class structure_file_format_vienna
         detail::consume(stream_view | view::take_until(!is_space));
 
         // make sure "buffer at end" implies "stream at end"
-        if ((std::istreambuf_iterator<char>{stream} == std::istreambuf_iterator<char>{}) && (!stream.eof()))
+        if ((stream_t{stream} == stream_t{})
+            && (!stream.eof()))
```
Comment (smehringer) [inline include/seqan3/io/structure_file/format_vienna.hpp]: can now go into line above :nail_care:

Comment (joergi-w) [reply include/seqan3/io/structure_file/format_vienna.hpp]: sure ;)

**Comment by joergi-w** [comment]
rebased on master

File: include/seqan3/io/structure_file/format_vienna.hpp [hunk 6]
```diff
@@ -129,9 +129,9 @@ class structure_file_format_vienna
               comment_type & SEQAN3_DOXYGEN_ONLY(comment),
               offset_type & SEQAN3_DOXYGEN_ONLY(offset))
     {
-        auto stream_view = view::subrange<decltype(std::istreambuf_iterator<char>{stream}),
-                                          decltype(std::istreambuf_iterator<char>{})>
-                           { std::istreambuf_iterator<char>{stream}, std::istreambuf_iterator<char>{} };
+        using stream_t = std::istreambuf_iterator<typename stream_type::char_type>;
```
Comment (h-2) [inline include/seqan3/io/structure_file/format_vienna.hpp]: It's not the stream type but the stream iterator type, maybe call it `stream_it_t`?

File: include/seqan3/io/structure_file/input.hpp [hunk 7]
```diff
@@ -809,45 +810,33 @@ class structure_file_in
...
+                      selected_field_ids const & SEQAN3_DOXYGEN_ONLY(fields_tag) = selected_field_ids{}) :
+        primary_stream{new std::ifstream{filename, std::ios_base::in | std::ios::binary}, stream_deleter_default}
     {
-        // open stream
-        stream.open(_file_name, std::ios_base::in | std::ios::binary);
-        if (!stream.is_open())
+        if (!primary_stream->good())
             throw file_open_error{"Could not open file for reading."};
 
+        // possibly add intermediate compression stream
```
Comment (h-2) [inline include/seqan3/io/structure_file/input.hpp]: **de**compression stream :nail_care:

File: include/seqan3/io/structure_file/input.hpp [hunk 8]
```diff
@@ -856,19 +845,47 @@ class structure_file_in
...
+    structure_file_in(stream_t & stream,
                       file_format const & SEQAN3_DOXYGEN_ONLY(format_tag),
                       selected_field_ids const & SEQAN3_DOXYGEN_ONLY(fields_tag) = selected_field_ids{}) :
-        stream{std::move(_stream)}, format{file_format{}}
+        primary_stream{&stream, stream_deleter_noop}, format{file_format{}}
     {
         static_assert(meta::in<valid_formats, file_format>::value,
                       "You selected a format that is not in the valid_formats of this file.");
 
+        // possibly add intermediate compression stream
```
Comment (h-2) [inline include/seqan3/io/structure_file/input.hpp]: **de**compression

File: include/seqan3/io/structure_file/format_vienna.hpp (hunk 6 above)
Comment (joergi-w) [reply include/seqan3/io/structure_file/format_vienna.hpp]: yes, good point!

File: include/seqan3/io/structure_file/input.hpp (hunk 7 above)
Comment (joergi-w) [reply include/seqan3/io/structure_file/input.hpp]: oups...

File: include/seqan3/io/structure_file/format_vienna.hpp [hunk 9]
```diff
@@ -129,9 +129,9 @@ class structure_file_format_vienna
               comment_type & SEQAN3_DOXYGEN_ONLY(comment),
               offset_type & SEQAN3_DOXYGEN_ONLY(offset))
     {
-        auto stream_view = view::subrange<decltype(std::istreambuf_iterator<char>{stream}),
-                                          decltype(std::istreambuf_iterator<char>{})>
-                           { std::istreambuf_iterator<char>{stream}, std::istreambuf_iterator<char>{} };
+        using stream_it_t = std::istreambuf_iterator<typename stream_type::char_type>;
+        auto stream_view = view::subrange<decltype(stream_it_t{stream}), decltype(stream_it_t{})>{stream_it_t{stream},
```
Comment (h-2) [inline include/seqan3/io/structure_file/format_vienna.hpp]: instead of writing `decltype(stream_it_t{stream})` you can just write `stream_it_t` :smile:

Comment (joergi-w) [reply include/seqan3/io/structure_file/format_vienna.hpp]: okay!

**Comment by h-2** [comment]
Has conflicts.

**Comment by joergi-w** [comment]
rebased on master

**Comment by h-2** [comment]
!rebuild

**Pull request merged**